[WIP] Bug/381 sortupdate event index wrong

mfeherpataky opened the pull request:
Following conversation in #381, also fixes #281.

Tested with linked lists, also with copy-only lists.
Not sure if all the cases were covered, so it needs some more manual tests.

--

FYI - Automatic tests are failing for ghost handling. 
Not sure if this was after a rebase on the new master, but I suspect it was as no ghost code was altered.

Unfortunately for the next few days will not have any free time to continue.

TL;DR - needs manual tests and if works as expected - update to documentation & automated tests code

mfeherpataky committed:
Quick fix to wrong sortable index - local cache of sortable items is refreshed on dragstart event

Custom Event results update - fixes #281 & #381

Code comments

Prevent processing of non-sortable elements on dragover

Logical flip of an expected boolean return

lukasoppermann commented:
I did not have time to test it yet, but having a look at the code it looks very good. 👍 
I am currently using `sortableContainer` for what you refer to as `sortableList` in the serialize function. I think Container might be better, because you can technically sort tables and grids as well. What do you think?

I think we need to add some development notes in some files and and link those in the readme, which should include naming, etc. I will start on this but if you have ideas for what you think would make working on it easier, please let me know. For e.g. concerning the setup and typescript.

mfeherpataky commented:
Yeap, I will refer to `sortableContainers` - that makes sense, especially as that's how we are naming them in the event responses.

I will come back to in on the weekend.
Did you find anything else I should fix/look into for this PR?

edit: Also - I will use `dragenter`, instead of `dragover` to fetch `itemsBeforeUpdate` for destinationContainer. The former is triggered only once and the latter every couple 100ms, with the same result. Just read specs on MDN 😄.

mfeherpataky commented on src/html5sortable.ts:
```diff
@@ -324,37 +339,58 @@ export default function sortable (sortableElements, options: object|string|undef
+    /*
+     We are capturing targetSortable data with 'dragover' event
+    */
+    _on(sortableElement, 'dragover', (e) => {
```
That will be changed to dragenter - much cheaper to run

lukasoppermann commented:
Hey, good catch! 👍 
The only thing apart from the `dragenter`, that I can think of, is updating the `readme` and maybe improving the examples so that testing this is easier.

I did check it out and it worked quite well in my test, very neat.

mfeherpataky committed:
README update, along with use of dragenter

mfeherpataky committed:
Add updated Docs in HTML and js to run docs

mfeherpataky commented:
hi @lukasoppermann 

Things done:
1. Rebased onto the latest master,
2. Renamed local variables to sortableContainer - from sortableList
3. Using `dragenter` to fire off less events when getting targetContainer itemsBeforeUpdate
4. Fixed an issue, where in the originContainer the itemsBeforeUpdate still carried placeholder
5. Updated README and docs/index.html to use new event object in console.log

--

However, I am a bit at a loss now with the tests results.
There are some local errors I get (around Ghost and Dragging) that never show up in Travis and vice-versa (export/import issues). To be honest, I cannot figure the tests out and could not since the beginning.

EDIT: Tests are passing now, though the early error messages in Travis had little - if anything - to do with the code I changed 😕.

mfeherpataky committed:
Added a missing removal of placeholder onDrop outside of a valid target & and unstaged /docs/html5sortable.js - until it is release ready

mfeherpataky committed:
Removing unnecessary code after a rebase to the latest master

mfeherpataky committed:
Making Travis happy - removing unnecessary whitespaces from around parameters passed on to a function

Commit:
Merge branch 'master' into bug/381-sortupdate-event-index-wrong

mfeherpataky commented:
I found still one issue with this branch - if `dragend` is called second (in cases of a successful drop) then it still attempts to remove the placeholder. Since it is not checking if the placeholder is still there, console errors out on calling .remove() on undefined.

Will fix it and push again

mfeherpataky committed:
Only attempt to remove placeholder if it IS there

lukasoppermann commented:
Very nice, thank you for the great work on this. 😃 This improves the events quite a lot!
